### PR TITLE
[14.0][ADD] partner_kind

### PR DESCRIPTION
**Pull request by syera94**
@sebastienbeau 

Here the partner_kind module to be able to select a type de partner.

### CODE REVIEW

File: partner_kind/i18n/fr.po [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#. module: partner_kind
+#: model:ir.model,name:partner_kind.model_res_users
+msgid "Users"
+msgstr "Utilisateurs"
+
+#. module: partner_kind
+#: model_terms:ir.ui.view,arch_db:partner_kind.view_partner_form
+#: model_terms:ir.ui.view,arch_db:partner_kind.view_res_partner_filter
+msgid "kind"
+msgstr "Type"
```
Comment (Kev-Roche) [inline partner_kind/i18n/fr.po:106]: The French translation need to be updated (for exemple "Individual", "Company")

File: partner_kind/models/res_partner.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    def _get_partner_kind(self):
+        return [
+            ("user", "User"),
+            ("address", "Address"),
+            ("person", "Individual"),
+            ("company", "Company"),
+        ]
+
+    @api.depends("user_ids", "parent_id", "company_type")
```
Comment (Kev-Roche) [inline partner_kind/models/res_partner.py]:
```suggestion
    @api.depends("user_ids", "parent_id", "is_company")
```

File: partner_kind/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    def _compute_kind(self):
+
+        for rec in self.with_context(active_test=False):
+            if rec.user_ids:
```
Comment (Kev-Roche) [inline partner_kind/models/res_partner.py]: With context active_test=False, the kind of an active partner will stay as "user" even if the related res.user was archived. 
I recommend the following:
```suggestion
    def _compute_kind(self):
        for rec in self:
          if any(rec.user_ids.mapped("active")):
```

also add "user_ids.active" in @api.depends

File: partner_kind/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                rec.kind = "user"
+            elif rec.parent_id:
+                rec.kind = "address"
+            elif rec.is_company is True:
```
Comment (Kev-Roche) [inline partner_kind/models/res_partner.py]:
```suggestion
            elif rec.is_company:
```

File: partner_kind/models/res_partner.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                if not rec.parent_id:
+                    rec.kind = "company"
+            elif rec.is_company is False:
```
Comment (Kev-Roche) [inline partner_kind/models/res_partner.py]:
```suggestion
            elif not rec.is_company:
```

File: partner_kind/models/res_partner.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+                if not rec.user_ids:
+                    rec.kind = "person"
+
+    @api.constrains("kind")
+    def _check_valid_kind(self):
+        if self._context.get("skip_kind_check"):
+            return
+        for rec in self.with_context(active_test=False):
+            if rec.kind == "user" and not rec.user_ids:
```
Comment (Kev-Roche) [inline partner_kind/models/res_partner.py]:
```suggestion
        for rec in self:
            if rec.kind == "user" and not any(rec.user_ids.mapped("active")):
```

File: partner_kind/models/res_partner.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    @api.depends("user_ids", "parent_id", "company_type")
+    def _compute_kind(self):
+
+        for rec in self.with_context(active_test=False):
+            if rec.user_ids:
+                rec.kind = "user"
+            elif rec.parent_id:
+                rec.kind = "address"
+            elif rec.is_company is True:
+                if not rec.parent_id:
```
Comment (Kev-Roche) [inline partner_kind/models/res_partner.py]:
```suggestion
```
also tested above

File: partner_kind/readme/DESCRIPTION.rst [hunk 8]
```diff
@@ -0,0 +1,16 @@
+This module add kind of partner to partners.
+In the method _get_partner_kind : there si the selection you want to add as partner kind. To mange this générique module, we add 4 selections which can obviously be removed.
+In this selection field you will find :
+- The recordset address : for the company or person linked to a parent
+- the recordset user : for company internal person which cannot be concerned by the kind
+- the recordset company : for the other case if it's a company
+- the recordset person : as the field is required, there is a default value
```
Comment (Kev-Roche) [inline partner_kind/readme/DESCRIPTION.rst]: They are not recordset, just selection options, please change it in all the readme. 
Maybe " - Address : for the company or person linked to a parent" is enough.
Also, you can use the interface term for users.

File: partner_kind/tests/test_partner_kind.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    def setUp(self):
+        super(TestPartnerKind, self).setUp()
+
+        self.Partner1 = self.env.ref("base.res_partner_address_15")
+        self.Partner2 = self.env.ref("base.partner_admin")
+        self.Partner3 = self.env.ref("base.res_partner_12")
+        self.Partner4 = self.env["res.partner"].create(
+            {
+                "name": "Test",
+                "company_type": "person",
```
Comment (Kev-Roche) [inline partner_kind/tests/test_partner_kind.py]: Not required as it's the default kind.

**Review by Kev-Roche — CHANGES_REQUESTED** [review]
Good job!  
Please rename the PR with the version number : [14.0][ADD] partner_kind
And add the readme authors part with you and @sebastienbeau in it.

File: partner_kind/models/res_partner.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                    rec.kind = "company"
+            elif rec.is_company is False:
+                if not rec.user_ids:
```
Comment (Kev-Roche) [inline partner_kind/models/res_partner.py]:
```suggestion
```
Already tested above

File: partner_kind/i18n/fr.po (hunk 1 above)
Comment (syera94) [reply partner_kind/i18n/fr.po:106]: we ll do together

File: partner_kind/models/res_partner.py (hunk 2 above)
Comment (syera94) [reply partner_kind/models/res_partner.py]: ok

File: partner_kind/models/res_partner.py (hunk 3 above)
Comment (syera94) [reply partner_kind/models/res_partner.py]: ok

File: partner_kind/models/res_partner.py (hunk 4 above)
Comment (syera94) [reply partner_kind/models/res_partner.py]: ok

File: partner_kind/models/res_partner.py (hunk 5 above)
Comment (syera94) [reply partner_kind/models/res_partner.py]: ok

File: partner_kind/models/res_partner.py (hunk 7 above)
Comment (syera94) [reply partner_kind/models/res_partner.py]: ok

File: partner_kind/readme/DESCRIPTION.rst (hunk 8 above)
Comment (syera94) [reply partner_kind/readme/DESCRIPTION.rst]: ok

File: partner_kind/tests/test_partner_kind.py (hunk 9 above)
Comment (syera94) [reply partner_kind/tests/test_partner_kind.py]: ok

File: partner_kind/models/res_partner.py (hunk 10 above)
Comment (syera94) [reply partner_kind/models/res_partner.py]: ok

File: partner_kind/readme/DESCRIPTION.rst [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+- Person : as the field is required, there is a default value
+
+The way to use this module:
+
+In the method _get_partner_kind, add the selection you need as partner kind.
+Once is done,if you want to remove all of the initial recordset, you can also remove the compute field or change it to fit better to your case
+if you only want to remove some of the initial recordset, you have to remove in the compute field the linked conditions
+After, the same thing as to be done in a contrains.
+
+For the tests, as they are linked to this specifiques initials recordsets, thet won't worked if you remove or change this initial recordsets.
```
Comment (Kev-Roche) [inline partner_kind/readme/DESCRIPTION.rst]: Duplicate of USAGE.srt. 
I'm not sure but I think that the readme needs to describe the module itself, not the potential changes that we can do on it.

File: partner_kind/readme/DESCRIPTION.rst [hunk 12]
```diff
@@ -0,0 +1,16 @@
+This module add kind of partner to partners.
```
Comment (Kev-Roche) [inline partner_kind/readme/DESCRIPTION.rst:1]: Check the typos on the file

**Review by Kev-Roche — CHANGES_REQUESTED** [review]

File: partner_kind/models/res_partner.py (hunk 3 above)
Comment (Kev-Roche) [reply partner_kind/models/res_partner.py]: we need both user_ids and user_ids.active

File: partner_kind/tests/test_partner_kind.py (hunk 9 above)
Comment (Kev-Roche) [reply partner_kind/tests/test_partner_kind.py]: only line 18 "company_type": "person", was unnecessary, you have to keep self.Partner4 and the related test.

File: partner_kind/models/res_partner.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+            ("address", "Address"),
+            ("person", "Individual"),
+            ("company", "Company"),
+        ]
+
+    @api.depends("user_ids", "user_ids.active", "parent_id", "is_company")
+    def _compute_kind(self):
+
+        for rec in self:
+            if rec.kind == "user" and not any(rec.user_ids.mapped("active")):
```
Comment (sebastienbeau) [inline partner_kind/models/res_partner.py]:
```
for res in self:
    rec._get_default_kind()
```

File: partner_kind/models/res_partner.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class ResPartner(models.Model):
+
+    _inherit = ["res.partner"]
+
```
Comment (sebastienbeau) [inline partner_kind/models/res_partner.py:11]: In the view hide the field "company_type"

```
is_company = fields.Boolean(                                    
    compute="_compute_is_company",
    compute_sudo=True,
    store=True,
)

def _get_is_company_kind(self):
    return ["company"]

@api.depends("kind")                      
def _compute_is_company(self):            
    for record in self:                   
         record.is_company =  record.kind in self._get_is_company_kind()      
```

File: partner_kind/models/res_partner.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        for rec in self:
+            if rec.kind == "user" and not any(rec.user_ids.mapped("active")):
+                rec.kind = "user"
+            elif rec.parent_id:
+                rec.kind = "address"
+            elif rec.is_company:
+                rec.kind = "company"
+            elif not rec.is_company:
+                rec.kind = "person"
```
Comment (sebastienbeau) [inline partner_kind/models/res_partner.py:45]: we can remove it if the is_company is a computed field ;)

**Comment by syera94** [comment]
@Kev-Roche : the changes which has been modified  since our conversation this morning (23/06/2023) --> as the field parent_id is linked to the field company_type, if i hide it, the field parent_id does not appear anymore. 
In my opinion, it's better to let it visible.
Check and see
@sebastienbeau FYI

File: partner_kind/models/res_partner.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    @api.depends("user_ids", "user_ids.active", "parent_id", "is_company")
+    def _compute_kind(self):
+        self._get_default_kind()
+
+    def _get_default_kind(self):
+        for rec in self:
+            if any(rec.user_ids.mapped("active")):
+                rec.kind = "user"
+            elif rec.parent_id:
+                rec.kind = "address"
```
Comment (paradoxxxzero) [inline partner_kind/models/res_partner.py:41]: A partner child can also be a contact hence a person

**Pull request closed without merging**